src/glyph_item_iter.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use glib::translate::*;
+use pango_sys;
+use std::marker::PhantomData;
+use GlyphItem;
+
+//Note: This type not exported
+glib_wrapper! {
+    #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct GlyphItemIter(Boxed<pango_sys::PangoGlyphItemIter>);
+
+    match fn {
+        copy => |ptr| pango_sys::pango_glyph_item_iter_copy(mut_override(ptr)),
+        free => |ptr| pango_sys::pango_glyph_item_iter_free(ptr),
+        init => |_ptr| (),
+        clear => |_ptr| (),
+        get_type => || pango_sys::pango_glyph_item_iter_get_type(),
+    }
+}
+
+impl GlyphItemIter {
+    pub(crate) unsafe fn init_end(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        let mut iter = GlyphItemIter::uninitialized();
+        let ret = from_glib(pango_sys::pango_glyph_item_iter_init_end(
+            iter.to_glib_none_mut().0,
+            mut_override(glyph_item.to_glib_none().0),
+            text.to_glib_none().0,
+        ));
+        if ret {
+            Some(iter)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) unsafe fn init_start(glyph_item: &GlyphItem, text: &str) -> Option<GlyphItemIter> {
+        let mut iter = GlyphItemIter::uninitialized();
+        let ret = from_glib(pango_sys::pango_glyph_item_iter_init_start(
+            iter.to_glib_none_mut().0,
+            mut_override(glyph_item.to_glib_none().0),
+            text.to_glib_none().0,
+        ));
+        if ret {
+            Some(iter)
+        } else {
+            None
+        }
+    }
+
+    pub fn next_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_next_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn prev_cluster(&mut self) -> bool {
+        unsafe {
+            from_glib(pango_sys::pango_glyph_item_iter_prev_cluster(
+                self.to_glib_none_mut().0,
+            ))
+        }
+    }
+
+    pub fn to_data(&self) -> GlyphItemIteratorData {
+        GlyphItemIteratorData {
+            start_glyph: self.0.start_glyph,
+            end_glyph: self.0.end_glyph,
+            start_index: self.0.start_index as usize,
+            end_index: self.0.end_index as usize,
+            start_char: self.0.start_char as usize,
+            end_char: self.0.end_char as usize,
+        }
+    }
+}
+
+pub struct GlyphItemIteratorData {
+    pub start_glyph: i32,

```

Comment:
Can this be indeed negative?

Reply:
https://gitlab.gnome.org/GNOME/pango/blob/050ebae3a0340e37187319273b5ab443cbe1f0fd/pango/pango-glyph-item.c#L481
but on my tests its minimum is 0

Reply:
That would be on RTL text, and I think the `pango_glyph_item_iter_prev_cluster()` call before returning would then set that to a positive index again. I don't think this can be negative ever.

Reply:
Ok, I change it too